src/lis3dh.toit

```diff
@@ -99,10 +125,66 @@ class Lis3dh:
     range_bits := range << 4
     high_resolution_bit := 0b1000
 
+    ctrl3 := 0x0  // Default value. Might be changed by free-fall detection below.
+
     ctrl4 := range_bits | high_resolution_bit
 
+    ctrl5 := 0x0 // Default value
+
+    int1_threshold := 0x0 // Default value
+    int1_duration := 0x0 // Default value
+    int1_cfg := 0x0 // Default value
+
+    if detect_free_fall:
+      // See chapter 6.4 of the sensor's application note:
+      // https://www.st.com/resource/en/application_note/an3308-lis3dh-mems-digital-output-motion-sensor-ultralowpower-highperformance-3axis-nano-accelerometer-stmicroelectronics.pdf
+
+      // Enable interrupt 1 to use the interrupt generator (int1_cfg).
+      ctrl3 = 0b100_0000  // IA1 interrupt on INT1.
+
+      // The threshold value depends on the range.
+      // The least significant bit in the register is worth:
+      // - 16mg @ 2G
+      // - 32mg @ 4G
+      // - 62mg @ 8G
+      // - 186mg @ 16G
+      // We want to detect free fall at roughly 350mg.
+      int1_threshold = range == RANGE_16G ? 2 : (0x16 >> range)
+
+      if free_fall_duration.in_ns <= 0: throw "INVALID_RANGE"
+
+      // Depending on the frequency of the chip, which is set in ctr1.
+      if rate == RATE_1HZ: int1_duration = free_fall_duration.in_s
+      else if rate == RATE_10HZ: int1_duration = free_fall_duration.in_ms / 100
+      else if rate == RATE_25HZ: int1_duration = free_fall_duration.in_ms / 40
+      else if rate == RATE_50HZ: int1_duration = free_fall_duration.in_ms / 20
+      else if rate == RATE_100HZ: int1_duration = free_fall_duration.in_ms / 10
+      else if rate == RATE_200HZ: int1_duration = free_fall_duration.in_ms / 5
+      else if rate == RATE_400HZ: int1_duration = (free_fall_duration * 4).in_ms / 10
+      else: unreachable
+
+      if int1_duration == 0: throw "Invalid rate/duration combination."
```

Comment:
We don't normally end exceptions with a '.'.

Reply:
done